src/lib/core/fiber_channel.c

```diff
@@ -191,7 +205,24 @@ fiber_channel_check_wait(struct fiber_channel *ch, ev_tstamp start_time,
 void
 fiber_channel_close(struct fiber_channel *ch)
 {
-	if (ch->is_closed)
+	if (!fiber_channel_graceful_close || ch->count == 0) {
+		/* Destroy if no graceful close or buffer is already empty. */
+		fiber_channel_destroy(ch);
+		return;
+	}
+	ch->is_closed = true;
+	/* Wake up all channel_put(), they will automatically cancel. */
+	struct fiber *f;
+	while (!rlist_empty(&ch->waiters)) {
+		f = rlist_first_entry(&ch->waiters, struct fiber, state);
+		fiber_channel_waiter_wakeup(f, FIBER_CHANNEL_WAIT_CLOSED);
+	}
```

Comment:
AFAIU if the buffer is not empty, the destructors for bufferred messages will not be called. It seems incorrect.

Let's add a test case that verifies that a destructor is called for a bufferred message at graceful close. It is okay to add only a unit test case here. But if you want to add a Lua test case, you can define a destructor this way: `ffi.gc(ffi.new('void *'), function(_self) <...> end)`. Don't forget to call `collectgarbage()` before check that the destructor is invoked.

Reply:
Argh, that's the point of graceful close. I need a fresh head to review.

Regarding the test case: it looks interesting anyway. However we should verify that the destructors are NOT called.

Reply:
We discussed it with Igor and he thinks that the test case is more about general guarantees of the fiber channel and it has no strict relation to this patchset. I'll close the discussion so.